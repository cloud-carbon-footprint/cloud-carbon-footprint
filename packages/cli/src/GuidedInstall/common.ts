/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */

import { promisify } from 'util'
import { exec as execCb } from 'child_process'
import dotenv from 'dotenv'
import fs from 'fs-extra'
import { confirm, input, list, prompt } from 'typed-prompts'
import { prop } from 'ramda'

const exec = promisify(execCb)
const stat = promisify(fs.stat)

export const log = (m: string, leading = true): void =>
  console.log(`${leading ? '\n' : ''}# ${m}\n`)

export const lineBreak = (): void => console.log()

export const listPrompt = (
  message: string,
  options: string[],
): Promise<string> =>
  prompt<{ key: string }>([list('key', message, options)]).then(prop('key'))

export const confirmPrompt = (
  message: string,
  question = 'Is this step complete?',
  requireYes = true,
): Promise<boolean> =>
  prompt<{ key: boolean }>([confirm('key', message + `\n${question}`)])
    .then(prop('key'))
    .then((result) => {
      if (requireYes && !result) {
        log('Please try again when you have completed this step.')
        process.exit(0)
      }
      console.log()
      return result
    })

export const inputPrompt = (
  message: string,
  required = true,
): Promise<string> =>
  prompt<{ key: string }>([input('key', message)])
    .then(prop('key'))
    .then((result: string) => {
      if (required && !result) {
        log('Please enter a value.')
        return inputPrompt(message, required)
      }
      return result
    })

export const runCmd = async (cmd: string): void => {
  try {
    await exec(cmd)
  } catch (error) {
    process.stdout.write(error.stderr)
    process.stdout.write(error.stdout)
    throw new Error(`Could not execute command ${cmd}`)
  }
}
// will load existing .env and overwrite the given keys
export const createEnvFile = async (
  dir: string,
  env: { [key: string]: string },
): void => {
  const path = `${dir}.env`
  log(`Creating ${path}...`, false)
  const exists = await stat(path)
    .then(() => true)
    .catch(() => false)

  const existing = exists ? dotenv.config({ path }) : {}
  if (existing.error) {
    throw existing.error
  }

  const newEnv = { ...existing.parsed, ...env }
  await fs.writeFile(
    path,
    Object.entries(newEnv).reduce(
      (config, [k, v]) => `${config}${k}=${v}\n`,
      '',
    ),
  )
  log('...done', false)
}
