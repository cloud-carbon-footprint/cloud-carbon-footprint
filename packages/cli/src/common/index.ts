/*
 * Â© 2021 Thoughtworks, Inc.
 */

import { confirm, input, list } from 'typed-prompts'
import { prop } from 'ramda'
import {
  exit,
  lineBreak,
  log,
  prompt,
  readConfig,
  resolve,
  stat,
  writeFile,
} from './external'

export type EnvConfig = { [key: string]: string }

export const microsite = 'https://cloud-carbon-footprint.org'

export const listPrompt = (
  message: string,
  options: string[],
): Promise<string> =>
  prompt<{ key: string }>(list('key', message, options)).then(prop('key'))

export const confirmPrompt = (
  message: string,
  question = 'Is this step complete?',
  requireYes = true,
): Promise<boolean> =>
  prompt<{ key: boolean }>(confirm('key', message + `\n${question}`))
    .then(prop('key'))
    .then((result) => {
      if (requireYes && !result) {
        log('Please try again when you have completed this step.')
        exit()
      }
      lineBreak()
      return result
    })

export const inputPrompt = (
  message: string,
  required = true,
): Promise<string> =>
  prompt<{ key: string }>(input('key', message))
    .then(prop('key'))
    .then((result: string) => {
      if (required && !result) {
        log('Please enter a value.')
        return inputPrompt(message, required)
      }
      return result
    })

export const createEnvFile = async (
  dir: string,
  env: { [key: string]: string },
): Promise<void> => {
  const path = resolve(`${dir}.env`)
  log(`Creating/updating ${path}...`, false)
  const exists = await stat(path)
    .then(() => true)
    .catch(() => false)

  const existing = exists ? readConfig(path) : { parsed: undefined }
  if (existing.error) {
    throw existing.error
  }

  const newEnv = { ...existing.parsed, ...env }
  await writeFile(
    path,
    Object.entries(newEnv).reduce(
      (config, [k, v]) => `${config}${k}=${v}\n`,
      '',
    ),
  )
  log('...done', false)
}
