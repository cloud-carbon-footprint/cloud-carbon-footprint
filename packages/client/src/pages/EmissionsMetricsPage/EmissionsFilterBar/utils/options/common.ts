/*
 * Â© 2021 Thoughtworks, Inc.
 */

import { DropdownOption } from 'Types'

export function isDropdownOptionInDropdownOptions(
  comparingDropdownOptions: DropdownOption[],
  dropdownOption: DropdownOption,
): boolean {
  let isWithinComparingDropdownOption = false
  comparingDropdownOptions.forEach((comparingDropdownOption) => {
    if (comparingDropdownOption.key === dropdownOption.key) {
      isWithinComparingDropdownOption = true
    }
  })
  return isWithinComparingDropdownOption
}
