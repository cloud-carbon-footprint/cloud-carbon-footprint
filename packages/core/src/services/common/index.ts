export { getPhysicalChips, calculateGigabyteHours } from './helpers'
export { QUERY_DATE_TYPES } from './types'
