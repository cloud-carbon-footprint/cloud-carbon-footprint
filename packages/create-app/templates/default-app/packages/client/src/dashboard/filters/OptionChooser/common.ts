/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */

import { DropdownOption } from '../DropdownFilter'

export function isDropdownOptionInDropdownOptions(
  comparingDropdownOptions: DropdownOption[],
  dropdownOption: DropdownOption,
): boolean {
  let isWithinComparingDropdownOption = false
  comparingDropdownOptions.forEach((comparingDropdownOption) => {
    if (comparingDropdownOption.key === dropdownOption.key) {
      isWithinComparingDropdownOption = true
    }
  })
  return isWithinComparingDropdownOption
}
