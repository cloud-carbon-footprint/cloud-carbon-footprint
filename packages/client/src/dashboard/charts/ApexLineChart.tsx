/*
 * Â© 2021 ThoughtWorks, Inc.
 */

import React, { FunctionComponent, useEffect } from 'react'
import { renderToStaticMarkup } from 'react-dom/server'
import { useTheme } from '@material-ui/core/styles'
import { GetApp, PanTool, RotateLeft, ZoomIn } from '@material-ui/icons'
import { CustomTooltip } from './CustomTooltip'

import { getChartColors } from '../../themes'
import { sumServiceTotals, getMaxOfDataSeries } from '../transformData'
import { EstimationResult } from '@cloud-carbon-footprint/common'
import Chart from 'react-apexcharts'
import ApexCharts from 'apexcharts'
import * as _ from 'lodash'

export interface DateRange {
  min: Date | null
  max: Date | null
}

type LegendToggle = {
  [key: string]: boolean
}

const formatDateToTime = (timestamp: string | Date) =>
  timestamp instanceof Date
    ? timestamp.getTime()
    : new Date(timestamp).getTime()

export const sortByDate = (data: EstimationResult[]): EstimationResult[] => {
  return data.sort((a: EstimationResult, b: EstimationResult) => {
    return formatDateToTime(a.timestamp) - formatDateToTime(b.timestamp)
  })
}

export const filterBy = (
  data: EstimationResult[],
  range: DateRange,
  defaultRange: DateRange,
): EstimationResult[] => {
  if (!range.min || !range.max) return data
  if (_.isEqual(range, defaultRange)) return data

  const min = formatDateToTime(range.min)
  const max = formatDateToTime(range.max)
  return data.filter((a: EstimationResult) => {
    const result = formatDateToTime(a.timestamp)
    return result >= min && result <= max
  })
}

export const ApexLineChart: FunctionComponent<ApexLineChartProps> = ({
  data,
}) => {
  const theme = useTheme()
  const [dateRange, setDateRange] = React.useState<DateRange>({
    min: null,
    max: null,
  })
  const [chartData, setChartData] = React.useState<EstimationResult[]>([])
  const [defaultRange, setDefaultRange] = React.useState<DateRange>({
    min: null,
    max: null,
  })
  const [toggledSeries, setToggledSeries] = React.useState<LegendToggle[]>([
    { CO2e: true },
    { 'Kilowatt Hours': false },
    { Cost: false },
  ])

  useEffect(() => {
    const newSortedData = sortByDate(data)
    const newDefaultRange = {
      min: newSortedData[0]?.timestamp
        ? new Date(newSortedData[0]?.timestamp)
        : null,
      max: newSortedData[newSortedData.length - 1]?.timestamp
        ? new Date(newSortedData[newSortedData.length - 1]?.timestamp)
        : null,
    }

    if (!_.isEqual(chartData, newSortedData)) setChartData(newSortedData)
    if (
      newDefaultRange.min instanceof Date &&
      newDefaultRange.max instanceof Date
    ) {
      if (!_.isEqual(defaultRange, newDefaultRange)) {
        setDateRange(newDefaultRange)
        setDefaultRange(newDefaultRange)
      }
    }
  }, [data])

  const filteredByZoomRange = filterBy(chartData, dateRange, defaultRange)
  // We need to get the HTML string version of these icons since ApexCharts doesn't take in custom React components.
  // Why, you might ask? Don't ask me, ask ApexCharts.
  const GetAppIconHTML = renderToStaticMarkup(<GetApp />)
  const PanToolIconHTML = renderToStaticMarkup(<PanTool />)
  const RotateLeftIconHTML = renderToStaticMarkup(<RotateLeft />)
  const ZoomInIconHTML = renderToStaticMarkup(<ZoomIn />)

  const cloudEstimationData = sumServiceTotals(filteredByZoomRange)
  const co2SeriesData = cloudEstimationData.co2Series
  const kilowattHoursSeriesData = cloudEstimationData.kilowattHoursSeries
  const costSeriesData = cloudEstimationData.costSeries
  const maxCO2e = getMaxOfDataSeries(co2SeriesData)
  const maxKilowattHours = getMaxOfDataSeries(kilowattHoursSeriesData)
  const maxCost = getMaxOfDataSeries(costSeriesData)

  useEffect(() => {
    ApexCharts.exec('lineChart', 'updateSeries', [
      {
        name: 'CO2e',
        data: co2SeriesData,
      },
      {
        name: 'Kilowatt Hours',
        data: kilowattHoursSeriesData,
      },
      {
        name: 'Cost',
        data: costSeriesData,
      },
    ])

    toggledSeries.forEach((legendToggle: LegendToggle) => {
      const [seriesKey, toggleValue] = Object.entries(legendToggle)[0]
      toggleValue
        ? ApexCharts.exec('lineChart', 'showSeries', [seriesKey])
        : ApexCharts.exec('lineChart', 'hideSeries', [seriesKey])
    })
  }, [co2SeriesData, kilowattHoursSeriesData, costSeriesData, toggledSeries])

  const colors = getChartColors(theme)
  const [blue, yellow, green] = [colors[0], colors[5], colors[8]]

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const options: any = {
    chart: {
      events: {
        beforeZoom: (chart: unknown, { xaxis }: { xaxis: DateRange }) => {
          const newFilteredData = filterBy(data, xaxis, defaultRange)

          if (newFilteredData.length >= 2) {
            setDateRange(xaxis)
            return {
              xaxis,
            }
          }
          return {
            dateRange,
          }
        },
        beforeResetZoom: () => {
          setDateRange(defaultRange)
        },
        legendClick: (chart: unknown, seriesIndex: number) => {
          const [seriesKey, toggledValue] = Object.entries(
            toggledSeries[seriesIndex],
          )[0]
          const toggleCheck = toggledSeries.filter(
            (series) => !!Object.values(series)[0],
          )
          const newToggledSeries = [...toggledSeries]
          newToggledSeries[seriesIndex] = {
            [seriesKey]: !toggledValue,
          }
          setToggledSeries(newToggledSeries)
          if (
            toggleCheck.length === 1 &&
            Object.keys(toggleCheck[0])[0] === seriesKey
          )
            setToggledSeries(toggledSeries)
        },
      },
      id: 'lineChart',
      background: theme.palette.background.paper,
      toolbar: {
        tools: {
          download: GetAppIconHTML,
          zoom: ZoomInIconHTML,
          zoomin: false,
          zoomout: false,
          pan: PanToolIconHTML,
          reset: RotateLeftIconHTML,
        },
      },
    },
    colors: [blue, yellow, green],
    height: '500px',
    series: [
      {
        name: 'CO2e',
        data: co2SeriesData,
      },
      {
        name: 'Kilowatt Hours',
        data: kilowattHoursSeriesData,
      },
      {
        name: 'Cost',
        data: costSeriesData,
      },
    ],
    stroke: {
      width: 1,
    },
    theme: {
      mode: theme.palette.type,
    },
    tooltip: {
      shared: true,
      custom: function ({ dataPointIndex }: { dataPointIndex: number }) {
        return renderToStaticMarkup(
          <CustomTooltip
            data={co2SeriesData}
            dataPointIndex={dataPointIndex}
          />,
        )
      },
    },
    title: {
      text: 'Cloud Usage',
      offsetY: -8,
      style: {
        fontSize: '24px',
      },
    },
    xaxis: {
      type: 'datetime',
      title: {
        text: '',
        offsetY: 18,
        style: {
          fontSize: '15px',
        },
      },
    },
    yaxis: [
      {
        max: 1.1 * maxCO2e,
        title: {
          text: 'CO2e (metric tons)',
          offsetX: -8,
          style: {
            fontSize: '15px',
          },
        },
        tickAmount: 10,
        decimalsInFloat: 3,
      },
      {
        max: 1.1 * maxKilowattHours,
        title: {
          text: 'Kilowatt Hours (kWh)',
          opposite: -8,
          style: {
            fontSize: '15px',
            color: yellow,
          },
        },
        tickAmount: 10,
        decimalsInFloat: 2,
        opposite: true,
        axisBorder: {
          show: true,
          color: yellow,
        },
        axisTicks: {
          show: false,
        },
        showAlways: false,
      },
      {
        max: 1.1 * maxCost,
        title: {
          text: 'Cost ($)',
          offsetX: 6,
          style: {
            fontSize: '15px',
            color: green,
          },
        },
        tickAmount: 10,
        decimalsInFloat: 2,
        opposite: true,
        axisBorder: {
          show: true,
          color: green,
          offsetX: -5,
        },
        axisTicks: {
          show: false,
        },
        showAlways: false,
      },
    ],
    grid: {
      padding: {
        bottom: +15,
        right: +20,
      },
    },
  }
  return (
    <Chart
      aria-label="apex-line-chart"
      options={options}
      series={options.series}
      type="line"
      height={options.height}
    />
  )
}

type ApexLineChartProps = {
  data: EstimationResult[]
}
