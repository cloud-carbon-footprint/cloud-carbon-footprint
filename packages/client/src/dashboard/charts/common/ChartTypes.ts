/*
 * Â© 2021 ThoughtWorks, Inc.
 */

import { EstimationResult } from '@cloud-carbon-footprint/common'

export type ApexChartProps = {
  data: EstimationResult[]
  dataType: string
}
