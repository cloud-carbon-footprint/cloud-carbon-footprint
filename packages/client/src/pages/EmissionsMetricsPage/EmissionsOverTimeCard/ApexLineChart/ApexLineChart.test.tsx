/*
 * Â© 2021 Thoughtworks, Inc.
 */

import React from 'react'
import { act, create, ReactTestRenderer } from 'react-test-renderer'
import moment from 'moment'
import ApexCharts from 'apexcharts'
import Chart from 'react-apexcharts'
import { render } from '@testing-library/react'
import ApexLineChart from './ApexLineChart'
import { GroupBy, ServiceData } from '@cloud-carbon-footprint/common'

jest.mock('apexcharts')
jest.mock('../../../../utils/themes')

describe('ApexLineChart', () => {
  class EstimationResultBuilder {
    private timestamp = moment('2019-08-10T00:00:00.000Z').toDate()
    private serviceEstimates: ServiceData[] = [
      {
        cloudProvider: 'AWS',
        accountId: 'account-1-id',
        accountName: 'account-1',
        serviceName: 'ebs',
        kilowattHours: 2,
        co2e: 5,
        cost: 3,
        region: 'us-west-2',
      },
      {
        cloudProvider: 'AWS',
        accountId: 'account-1-id',
        accountName: 'account-1',
        serviceName: 's3',
        kilowattHours: 4,
        co2e: 10,
        cost: 6,
        region: 'us-west-2',
      },
      {
        cloudProvider: 'AWS',
        accountId: 'account-1-id',
        accountName: 'account-1',
        serviceName: 'ec2',
        kilowattHours: 6,
        co2e: 15,
        cost: 9,
        region: 'us-west-2',
      },
    ]

    withTime(time: Date) {
      this.timestamp = time
      return this
    }

    build() {
      return {
        timestamp: this.timestamp,
        periodStartDate: this.timestamp,
        periodEndDate: this.timestamp,
        serviceEstimates: this.serviceEstimates,
        groupBy: GroupBy.day,
      }
    }
  }

  const setDateRangeSpy = jest.fn()
  const setChartDataSpy = jest.fn()
  const setDefaultDateRangeSpy = jest.fn()
  const setToggledSeriesSpy = jest.fn()
  // Use this function to mock the component's state, and pass any initial values you'd want to override for a specific test
  const mockUseState = (
    initialDateRange = { min: null, max: null },
    initialChartData = [],
    initialDefaultDateRange = { min: null, max: null },
    initialToggledSeries = [
      { CO2e: true },
      { 'Kilowatt Hours': false },
      { Cost: false },
    ],
  ) => {
    jest
      .spyOn(React, 'useState')
      .mockReturnValueOnce([initialDateRange, setDateRangeSpy])
      .mockReturnValueOnce([initialChartData, setChartDataSpy])
      .mockReturnValueOnce([initialDefaultDateRange, setDefaultDateRangeSpy])
      .mockReturnValueOnce([initialToggledSeries, setToggledSeriesSpy])
  }

  it('renders with correct configuration', () => {
    const root = create(
      <ApexLineChart
        data={[
          new EstimationResultBuilder()
            .withTime(new Date('2019-08-10T00:00:00.000Z'))
            .build(),
        ]}
      />,
    )
    expect(root.toJSON()).toMatchSnapshot()
  })

  it('should manually disable kilowatt hours and cost series on initial data load', () => {
    render(
      <ApexLineChart
        data={[
          new EstimationResultBuilder()
            .withTime(new Date('2019-08-10T00:00:00.000Z'))
            .build(),
        ]}
      />,
    )
    expect(ApexCharts.exec).toHaveBeenCalledWith('lineChart', 'hideSeries', [
      'Cost',
    ])
    expect(ApexCharts.exec).toHaveBeenCalledWith('lineChart', 'hideSeries', [
      'Kilowatt Hours',
    ])
  })

  it('should update chart with new data and default max values on props data change', () => {
    const testRenderer = create(<ApexLineChart data={[]} />)
    act(() => {
      testRenderer.update(
        <ApexLineChart data={[new EstimationResultBuilder().build()]} />,
      )
    })
    const filteredSeries = testRenderer.root.findByType(Chart)?.props?.series
    expect(filteredSeries).toBeDefined()
    expect(filteredSeries).toEqual([
      {
        name: 'CO2e',
        data: [
          {
            x: new Date('2019-08-10T00:00:00.000Z'),
            y: 30,
            usesAverageCPUConstant: false,
            cost: 18,
            kilowattHours: 12,
          },
        ],
      },
      {
        name: 'Kilowatt Hours',
        data: [
          {
            x: new Date('2019-08-10T00:00:00.000Z'),
            y: 12,
          },
        ],
      },
      {
        name: 'Cost',
        data: [
          {
            x: new Date('2019-08-10T00:00:00.000Z'),
            y: 18,
          },
        ],
      },
    ])
  })

  it('should set date range when data actually provided', () => {
    mockUseState()

    act(() => {
      create(<ApexLineChart data={[new EstimationResultBuilder().build()]} />)
    })

    expect(setDateRangeSpy).toHaveBeenCalledTimes(1)
  })

  it('should not set date range state when zooming with apex line chart and less than two filtered items are within the range', () => {
    mockUseState()

    let testRenderer: ReactTestRenderer
    act(() => {
      testRenderer = create(<ApexLineChart data={[]} />)
    })

    act(() => {
      const beforeZoomCallback =
        testRenderer.root?.findByType(Chart)?.props?.options?.chart?.events
          ?.beforeZoom

      expect(beforeZoomCallback).toBeDefined()
      beforeZoomCallback(undefined, {
        xaxis: {
          min: new Date('2019-01-10T00:00:00.000Z'),
          max: new Date('2019-08-10T00:00:00.000Z'),
        },
      })
    })

    expect(setDateRangeSpy).not.toHaveBeenCalled()
  })

  it('should set date range state when zooming with apex line chart and at least two filtered items are within the range', () => {
    mockUseState()

    let testRenderer: ReactTestRenderer
    act(() => {
      testRenderer = create(
        <ApexLineChart
          data={[
            new EstimationResultBuilder()
              .withTime(new Date('2019-05-10T00:00:00.000Z'))
              .build(),
            new EstimationResultBuilder()
              .withTime(new Date('2019-06-10T00:00:00.000Z'))
              .build(),
          ]}
        />,
      )
    })

    act(() => {
      const beforeZoomCallback =
        testRenderer.root?.findByType(Chart)?.props?.options?.chart?.events
          ?.beforeZoom

      expect(beforeZoomCallback).toBeDefined()
      beforeZoomCallback(undefined, {
        xaxis: {
          min: new Date('2019-01-10T00:00:00.000Z'),
          max: new Date('2019-08-10T00:00:00.000Z'),
        },
      })
    })

    expect(setDateRangeSpy).toHaveBeenCalledWith({
      min: new Date('2019-01-10T00:00:00.000Z'),
      max: new Date('2019-08-10T00:00:00.000Z'),
    })
  })

  it('should set the current default range state when resetting zoom through apex charts', () => {
    mockUseState(undefined, undefined, {
      min: new Date('2019-07-10T00:00:00.000Z'),
      max: new Date('2019-11-10T00:00:00.000Z'),
    })

    let testRenderer: ReactTestRenderer
    act(() => {
      testRenderer = create(<ApexLineChart data={[]} />)
    })

    act(() => {
      const beforeResetZoomCallback =
        testRenderer.root?.findByType(Chart)?.props?.options?.chart?.events
          ?.beforeResetZoom

      expect(beforeResetZoomCallback).toBeDefined()
      beforeResetZoomCallback()
    })

    expect(setDateRangeSpy).toHaveBeenCalledTimes(1)
    expect(setDateRangeSpy).toHaveBeenLastCalledWith({
      min: new Date('2019-07-10T00:00:00.000Z'),
      max: new Date('2019-11-10T00:00:00.000Z'),
    })
  })

  it('should update data based on new ranges', () => {
    const newDateRange = {
      min: new Date('2019-07-10T00:00:00.000Z'),
      max: new Date('2019-11-10T00:00:00.000Z'),
    }
    const defaultDateRange = {
      min: new Date('2019-06-10T00:00:00.000Z'),
      max: new Date('2019-12-10T00:00:00.000Z'),
    }
    const initialData = [
      new EstimationResultBuilder()
        .withTime(new Date('2019-06-10T00:00:00.000Z'))
        .build(),
      new EstimationResultBuilder()
        .withTime(new Date('2019-08-10T00:00:00.000Z'))
        .build(),
      new EstimationResultBuilder()
        .withTime(new Date('2019-10-10T00:00:00.000Z'))
        .build(),
      new EstimationResultBuilder()
        .withTime(new Date('2019-12-10T00:00:00.000Z'))
        .build(),
    ]

    mockUseState(newDateRange, initialData, defaultDateRange)

    let testRenderer: ReactTestRenderer
    act(() => {
      testRenderer = create(<ApexLineChart data={initialData} />)
    })

    act(() => {
      const filteredSeries = testRenderer.root?.findByType(Chart)?.props?.series
      expect(filteredSeries).toBeDefined()
      expect(filteredSeries).toEqual([
        {
          name: 'CO2e',
          data: [
            {
              x: new Date('2019-08-10T00:00:00.000Z'),
              y: 30,
              usesAverageCPUConstant: false,
              cost: 18,
              kilowattHours: 12,
            },
            {
              x: new Date('2019-10-10T00:00:00.000Z'),
              y: 30,
              usesAverageCPUConstant: false,
              cost: 18,
              kilowattHours: 12,
            },
          ],
        },
        {
          name: 'Kilowatt Hours',
          data: [
            {
              x: new Date('2019-08-10T00:00:00.000Z'),
              y: 12,
            },
            {
              x: new Date('2019-10-10T00:00:00.000Z'),
              y: 12,
            },
          ],
        },
        {
          name: 'Cost',
          data: [
            {
              x: new Date('2019-08-10T00:00:00.000Z'),
              y: 18,
            },
            {
              x: new Date('2019-10-10T00:00:00.000Z'),
              y: 18,
            },
          ],
        },
      ])
    })
  })

  it('should update toggle series state when a series is clicked in the legend', () => {
    mockUseState()

    let testRenderer: ReactTestRenderer
    act(() => {
      testRenderer = create(<ApexLineChart data={[]} />)
    })

    act(() => {
      const afterLegendClickCallback =
        testRenderer.root?.findByType(Chart)?.props?.options?.chart?.events
          ?.legendClick

      expect(afterLegendClickCallback).toBeDefined()
      afterLegendClickCallback(undefined, 1)
    })

    expect(setToggledSeriesSpy).toHaveBeenCalledWith([
      { CO2e: true },
      { 'Kilowatt Hours': true },
      { Cost: false },
    ])
  })
})
