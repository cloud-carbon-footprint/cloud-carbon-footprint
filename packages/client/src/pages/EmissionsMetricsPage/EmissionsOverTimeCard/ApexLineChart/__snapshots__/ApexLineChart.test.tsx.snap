// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApexLineChart renders with correct configuration 1`] = `
<div
  aria-label="apex-line-chart"
/>
`;
